pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,225 @@
+//	  2.  Recycle - PV is scrubbed clean and the ClaimRef is removed (will match Pending above on sync)
+//	  3.  Delete - PV is removed from infrastructure and PV is deleted via API
+//	  4.  Failed - PV failed to Recycle or Delete.
+type persistentVolumeReleasedState struct{}
+
+// persistentVolumePendingState determines if a PersistentVolume should be Pending phase.
+// The PV is Pending when it is not bound (no ClaimRef) or it is bound (has ClaimRef) but is not fully provisioned.
+func (state *persistentVolumePendingState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	if ctx.volume.Spec.ClaimRef == nil {
+		return true
+	}
+	if isBeingProvisioned(ctx.volume) {
+		return true
+	}
+	return false
+}
+
+func (state *persistentVolumePendingState) processState(ctx *persistentVolumeStateContext) error {
+	if err := ctx.updatePersistentVolumeStatus(api.VolumePending); err != nil {
+		return err
+	}
+
+	// launch Provisioner in separate thread / Goroutine
+	// lock on pv.name like the main control, guaranteeing 1 thread processing
+	// a PV at a time but allowing concurrent processing of multiple volumes
+	// TODO: make this work :)
+	if !isProvisioningComplete(ctx.volume) {
+		// go provisionVolume(ctx.volume)
+	}
+
+	return nil
+}
+
+// persistentVolumeBoundState determines if a PersistentVolume should be in Bound phase.
+// The PV is Pending when it is not bound (no ClaimRef) or it is bound (has ClaimRef) but is not fully provisioned.
+func (state *persistentVolumeBoundState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	if ctx.volume.Spec.ClaimRef == nil {
+		return false
+	}
+	if isBeingProvisioned(ctx.volume) {
+		return false
+	}
+	// have a claimRef but 404 means user deleted the claim. no longer bound, needs recycling.
+	if ctx.claimNotFound {
+		return false
+	}
+	// have ClaimRef and completed (or don't need) provisioning
+	return true
+}
+
+func (state *persistentVolumeBoundState) processState(ctx *persistentVolumeStateContext) error {
+	if err := ctx.updatePersistentVolumeStatus(api.VolumeBound); err != nil {
+		return err
+	}
+	return nil
+}
+
+// persistentVolumeReleasedState determines if a PersistentVolume should be in Released phase.
+func (state *persistentVolumeReleasedState) isCurrentState(ctx *persistentVolumeStateContext) bool {
+	// have claimRef but claim 404'd, indicating user deleted it, unbinding this volume
+	if ctx.volume.Spec.ClaimRef != nil && ctx.claimNotFound {
+		return true
+	}
+	return false
+}
+
+func (state *persistentVolumeReleasedState) processState(ctx *persistentVolumeStateContext) error {
+	if err := ctx.updatePersistentVolumeStatus(api.VolumeReleased); err != nil {
+		return err
+	}
+
+	// launch Recycler in separate thread / Goroutine
+	// lock on pv.name like the main control, guaranteeing 1 thread processing
+	// a PV at a time but allowing concurrent processing of multiple volumes
+	// TODO: make this work :)
+	//go reclaimVolume(ctx.volume)
```

Comment:
Just a note for future implementation: we do not have any annotation 'isBeingRecycled', so the implementation must *somehow* make sure that we don't start multiple recyclers/deleters for one PV.